doc/barclamp_development_exercises.md

```diff
@@ -0,0 +1,341 @@
+
+```
+[DEFAULT]
+debug = true
+insecure_debug = true
+```
+
+Restart the keystone service:
+
+```
+# systemctl restart apache2
+```
+
+The ``insecure_debug`` option is now enabled in keystone, which Crowbar doesn't
+support. This option is useful for debugging authentication requests on
+non-production systems. You can see it in action by requesting tokens from
+keystone using invalidate credentials or incorrect scopes.
+
+## Adding a Parameter to a Barclamp
+
+Adding a new parameter involves making changes to several components, but is
+ultimately a matter of copying and pasting similar examples. This exercise will
+add the ``insecure_debug`` parameter to the keystone barclamp. In reality we
+should never support setting this option through Crowbar because it is insecure
+for production use, but it is useful as an exercise.
+
+### Exercise
+
+This is done in three stages.
+
+#### Add the parameter to the data bag
+
+Edit the keystone data bag schema in
+``/opt/dell/chef/data_bags/crowbar/template-keystone.schema``. Add a new
+parameter ``insecure_debug`` under the existing ``debug`` parameter:
+
+```
+{
+  "type": "map", "required": true,
+  "mapping": {
+    "id": { "type": "str", "required": true, "pattern": "/^keystone-|^template-keystone$/" },
+    "description": { "type": "str", "required": true },
+    "attributes": { "type": "map", "required": true,
+      "mapping": {
+        "keystone": { "type": "map", "required": true,
+             "mapping": {
+                    "debug": { "type": "bool", "required": true },
+                    "insecure_debug": { "type": "bool", "required": true },
+...
+```
+Edit the default data bag in
+``/opt/dell/chef/data_bags/crowbar/template-keystone.json``. Add the new
+parameter there as well. Here you should also increment the schema revision
+number.
+
+```
+{
+  "id": "template-keystone",
+  "description": "Centralized authentication and authorization service for OpenStack",
+  "attributes": {
+    "keystone": {
+      "debug": false,
+      "insecure_debug": false,
+...
+  },
+  "deployment": {
+    "keystone": {
+      "crowbar-revision": 0,
+      "crowbar-applied": false,
+      "schema-revision": 302,
+```
+
+Add a schema migration in ``/opt/dell/chef/data_bags/migrate/keystone/``. The
+``300_noop.rb`` migration can be copied for the function prototypes. The file must
+start with the new schema revision number that you incremented in the default
+data bag and should describe the migration action. For example, if the new
+revision number is 302, the schema migration should go in
+``/opt/dell/chef/data_bags/migrate/keystone/302_add_insecure_debug.rb``. The
+upgrade migration needs to take the new parameter from the template attributes
+hash and add it to the attributes hash, and the downgrade needs to do the
+opposite. It will look something like this:
+
+```
+def upgrade(template_attrs, template_deployment, attrs, deployment)
+  attrs["insecure_debug"] = template_attrs["insecure_debug"]
```

Comment:
Please add a check if the value is already present - that's especially important for the upgrades, as it could happen that the value that actually already exists is rewritten by the empty one from the template. And even experienced developers make that mistake...